core/src/main/java/ai/timefold/solver/core/impl/move/streams/maybeapi/generic/provider/ListChangeMoveProvider.java

```diff
@@ -0,0 +1,118 @@
+package ai.timefold.solver.core.impl.move.streams.maybeapi.generic.provider;
+
+import java.util.Objects;
+
+import ai.timefold.solver.core.impl.move.streams.maybeapi.BiDataFilter;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.DataJoiners;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.generic.move.ListAssignMove;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.generic.move.ListChangeMove;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.generic.move.ListUnassignMove;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.stream.MoveProducer;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.stream.MoveProvider;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.stream.MoveStreamFactory;
+import ai.timefold.solver.core.preview.api.domain.metamodel.ElementPosition;
+import ai.timefold.solver.core.preview.api.domain.metamodel.PlanningListVariableMetaModel;
+import ai.timefold.solver.core.preview.api.domain.metamodel.PositionInList;
+import ai.timefold.solver.core.preview.api.domain.metamodel.UnassignedElement;
+
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public class ListChangeMoveProvider<Solution_, Entity_, Value_>
+        implements MoveProvider<Solution_> {
+
+    private final PlanningListVariableMetaModel<Solution_, Entity_, Value_> variableMetaModel;
+    private final BiDataFilter<Solution_, Entity_, Value_> isValueInListFilter;
+    private final BiDataFilter<Solution_, Value_, ElementPosition> validChangeFilter;
+
+    public ListChangeMoveProvider(PlanningListVariableMetaModel<Solution_, Entity_, Value_> variableMetaModel) {
+        this.variableMetaModel = Objects.requireNonNull(variableMetaModel);
+        this.isValueInListFilter = (solution, entity, value) -> {
+            if (entity == null || value == null) {
```

Comment:
This condition bothers me. Can the entity or value be null? Is there an alternative method to verify if all dependencies are complete?

Reply:
After reviewing the move generation logic, the comment is now clearer and makes more sense.